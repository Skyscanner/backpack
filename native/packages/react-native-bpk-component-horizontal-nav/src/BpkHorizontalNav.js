/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ScrollView, StyleSheet, View, ViewPropTypes } from 'react-native';
import React from 'react';
import PropTypes from 'prop-types';
import {
  colorGray100,
  borderSizeSm,
} from 'bpk-tokens/tokens/base.react.native';

import BpkHorizontalNavSelectedIndicator from './BpkHorizontalNavSelectedIndicator';
import withAnimatedProps from './withAnimatedProps';

const AnimatedIndicator = withAnimatedProps(BpkHorizontalNavSelectedIndicator, [
  'xOffset',
  'width',
]);

const styles = StyleSheet.create({
  nav: {
    borderColor: 'transparent',
    borderBottomColor: colorGray100,
    flexDirection: 'column',
    borderBottomWidth: borderSizeSm,
  },
  // to allow ScrollView item to be properly positioned in the spaceAround config
  // the contentContainerStyle needs flexGrow to have the children define the end
  // width and allow overflow from the view
  navSpaceAround: {
    flexGrow: 1,
  },
  inner: {
    flexDirection: 'row',
  },
  innerSpaceAround: {
    justifyContent: 'space-around',
  },
});

class BpkHorizontalNav extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      indicatorOffsetX: null,
      indicatorWidth: null,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.selectedId && this.childrenPositions[nextProps.selectedId]) {
      const nextLayoutProps = this.childrenPositions[nextProps.selectedId];
      this.setState({
        indicatorOffsetX: nextLayoutProps.x,
        indicatorWidth: nextLayoutProps.width,
      });
    }
  }

  onChildLayout = (event, id) => {
    const { width, x } = event.nativeEvent.layout;
    this.childrenPositions[id] = { width, x };

    // If the child in question is the initially selected one, the indicator can now be positioned.
    if (this.props.selectedId === id) {
      this.setState({
        indicatorOffsetX: x,
        indicatorWidth: width,
      });
    }
  };

  childrenPositions = {};

  render() {
    const { children, selectedId, spaceAround, style, ...rest } = this.props;

    const navStyle = [styles.nav];
    const innerViewStyle = [styles.inner];

    if (spaceAround) {
      navStyle.push(styles.navSpaceAround);
      innerViewStyle.push(styles.innerSpaceAround);
    }

    if (style) {
      navStyle.push(style);
    }

    const enhancedChildren = React.Children.map(
      children,
      child =>
        React.isValidElement(child) &&
        React.cloneElement(child, {
          key: child.props.id,
          selected: selectedId === child.props.id,

          // Have children report their layout details after being laid out.
          // This allows the selected indicator to be correctly positioned.
          onLayout: event => this.onChildLayout(event, child.props.id),
        }),
    );

    const renderIndicator = () => {
      if (this.state.indicatorOffsetX === null) {
        return null;
      }
      return (
        <AnimatedIndicator
          xOffset={this.state.indicatorOffsetX}
          width={this.state.indicatorWidth}
        />
      );
    };

    return (
      <ScrollView
        alwaysBounceHorizontal={false}
        contentContainerStyle={navStyle}
        horizontal
        showsHorizontalScrollIndicator={false}
        {...rest}
      >
        <View style={innerViewStyle}>{enhancedChildren}</View>
        {renderIndicator()}
      </ScrollView>
    );
  }
}

BpkHorizontalNav.propTypes = {
  children: PropTypes.node.isRequired,
  selectedId: PropTypes.string.isRequired,
  spaceAround: PropTypes.bool,
  style: ViewPropTypes.style,
};

BpkHorizontalNav.defaultProps = {
  spaceAround: false,
  style: null,
};

export default BpkHorizontalNav;
