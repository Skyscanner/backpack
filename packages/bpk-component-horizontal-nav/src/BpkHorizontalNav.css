@keyframes bpk-keyframe-spin{100%{transform:rotate(1turn)}}.bpk-horizontal-nav{box-shadow:0 1px 0 0 #dddde5 }.bpk-horizontal-nav__list{display:flex;margin:0;padding:0;align-items:stretch;list-style-type:none}
