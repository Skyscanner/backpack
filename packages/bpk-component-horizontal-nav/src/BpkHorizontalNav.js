/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/* @flow strict */

import PropTypes from 'prop-types';
import React, { Component, type Node } from 'react';

import { cssModules } from '../../bpk-react-utils';
import BpkMobileScrollContainer from '../../bpk-component-mobile-scroll-container';

import STYLES from './BpkHorizontalNav.module.scss';

const getClassName = cssModules(STYLES);

const HORIZONTAL_NAV_TYPES = {
  default: 'default',
  light: 'light',
};

export type Props = {
  autoScrollToSelected: boolean,
  children: Node,
  showUnderline: boolean,
  type: $Keys<typeof HORIZONTAL_NAV_TYPES>,
  className: ?string,
  leadingScrollIndicatorClassName: ?string,
  trailingScrollIndicatorClassName: ?string,
};

const getPos = (ref: ?Element): ?{ left: number, right: number } => {
  if (!ref) {
    return null;
  }
  const pos = ref.getBoundingClientRect();
  return pos;
};

class BpkHorizontalNav extends Component<Props> {
  selectedItemRef: ?Element;

  scrollRef: ?Element;

  static propTypes = {
    children: PropTypes.node.isRequired,
    autoScrollToSelected: PropTypes.bool,
    className: PropTypes.string,
    leadingScrollIndicatorClassName: PropTypes.string,
    showUnderline: PropTypes.bool,
    trailingScrollIndicatorClassName: PropTypes.string,
    type: PropTypes.oneOf(Object.keys(HORIZONTAL_NAV_TYPES)),
  };

  static defaultProps = {
    autoScrollToSelected: false,
    className: null,
    leadingScrollIndicatorClassName: null,
    showUnderline: true,
    trailingScrollIndicatorClassName: null,
    type: HORIZONTAL_NAV_TYPES.default,
  };

  constructor(props: Props) {
    super(props);

    this.scrollRef = null;
    this.selectedItemRef = null;
  }

  componentDidMount = () => {
    this.scrollSelectedIntoView(false);
  };

  componentDidUpdate() {
    this.scrollSelectedIntoView(true);
  }

  scrollSelectedIntoView = (useSmoothScroll: boolean) => {
    if (
      !this.props.autoScrollToSelected ||
      !this.scrollRef ||
      !this.selectedItemRef
    ) {
      return;
    }

    if (!this.scrollRef) {
      return;
    }

    const selectedItemPos = getPos(this.selectedItemRef);
    if (!selectedItemPos) {
      return;
    }

    const scrollPos = getPos(this.scrollRef);
    if (!scrollPos) {
      return;
    }

    // We only need to carry out a scroll if the element is not already fully within the scroll view
    const needsScroll =
      selectedItemPos.left < scrollPos.left ||
      selectedItemPos.right > scrollPos.right;

    if (!needsScroll) {
      return;
    }

    const scrollAdjustment = selectedItemPos.left - scrollPos.left;
    // Some browsers don't support smooth scrolling, so in those cases we must fall back to simply setting `scrollLeft`
    if (
      this.scrollRef &&
      this.scrollRef.scroll &&
      typeof this.scrollRef.scroll === 'function' &&
      useSmoothScroll
    ) {
      this.scrollRef.scroll({
        left: scrollAdjustment,
        behavior: 'smooth',
      });
    } else if (this.scrollRef) {
      this.scrollRef.scrollLeft = scrollAdjustment;
    }
  };

  render() {
    const {
      autoScrollToSelected,
      children: rawChildren,
      className,
      leadingScrollIndicatorClassName,
      showUnderline,
      trailingScrollIndicatorClassName,
      type,
      ...rest
    } = this.props;

    const classNames = getClassName(
      'bpk-horizontal-nav',
      showUnderline && `bpk-horizontal-nav--show-${type}-underline`,
      className,
    );

    let children = rawChildren;

    if (autoScrollToSelected || type === HORIZONTAL_NAV_TYPES.light) {
      children = React.Children.map(rawChildren, (child) => {
        const childProps = {};
        let childRef;
        if (autoScrollToSelected) {
          if (child && child.props && child.props.selected) {
            childRef = (ref) => {
              this.selectedItemRef = ref;
            };
          }
        }

        if (type === HORIZONTAL_NAV_TYPES.light) {
          childProps.type = HORIZONTAL_NAV_TYPES.light;
        }

        return child ? (
          <div ref={childRef}>{React.cloneElement(child, childProps)}</div>
        ) : null;
      });
    }

    return (
      // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
      <BpkMobileScrollContainer
        innerContainerTagName="nav"
        className={classNames}
        leadingIndicatorClassName={leadingScrollIndicatorClassName}
        trailingIndicatorClassName={trailingScrollIndicatorClassName}
        scrollerRef={(ref) => {
          this.scrollRef = ref;
        }}
        {...rest}
      >
        <div
          className={getClassName('bpk-horizontal-nav__list')}
          role="tablist"
        >
          {children}
        </div>
      </BpkMobileScrollContainer>
    );
  }
}

export default BpkHorizontalNav;
export { HORIZONTAL_NAV_TYPES };
