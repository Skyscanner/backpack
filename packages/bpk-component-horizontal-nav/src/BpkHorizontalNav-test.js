/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2020 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import renderer from 'react-test-renderer';

import BpkHorizontalNav from './BpkHorizontalNav';

describe('BpkHorizontalNav', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(<BpkHorizontalNav>My nav content.</BpkHorizontalNav>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with custom "className" prop', () => {
    const tree = renderer
      .create(
        <BpkHorizontalNav className="my-custom-class-name">
          My nav content.
        </BpkHorizontalNav>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with custom "leadingScrollIndicatorClassName" prop', () => {
    const tree = renderer
      .create(
        <BpkHorizontalNav leadingScrollIndicatorClassName="my-custom-class-name">
          My nav content.
        </BpkHorizontalNav>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with custom "trailingScrollIndicatorClassName" prop', () => {
    const tree = renderer
      .create(
        <BpkHorizontalNav trailingScrollIndicatorClassName="my-custom-class-name">
          My nav content.
        </BpkHorizontalNav>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with arbitrary props', () => {
    const tree = renderer
      .create(
        <BpkHorizontalNav
          data-arbitrary-1="arbirary-value-1"
          data-arbitrary-2="arbirary-value-2"
        >
          My nav content.
        </BpkHorizontalNav>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
