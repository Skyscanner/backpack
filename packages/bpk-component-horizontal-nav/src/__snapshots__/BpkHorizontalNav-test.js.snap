// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkHorizontalNav should render correctly 1`] = `
<DocumentFragment>
  <div
    class="bpk-mobile-scroll-container bpk-horizontal-nav bpk-horizontal-nav--show-default-underline"
    style="height: auto;"
  >
    <div
      class="bpk-mobile-scroll-container__scroller"
    >
      <nav
        class="bpk-mobile-scroll-container__inner"
      >
        <div
          class="bpk-horizontal-nav__list"
          role="tablist"
        >
          My nav content.
        </div>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BpkHorizontalNav should render correctly with "showUnderline" set to false 1`] = `
<DocumentFragment>
  <div
    class="bpk-mobile-scroll-container bpk-horizontal-nav"
    style="height: auto;"
  >
    <div
      class="bpk-mobile-scroll-container__scroller"
    >
      <nav
        class="bpk-mobile-scroll-container__inner"
      >
        <div
          class="bpk-horizontal-nav__list"
          role="tablist"
        >
          My nav content.
        </div>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BpkHorizontalNav should render correctly with "type" prop 1`] = `
<DocumentFragment>
  <div
    class="bpk-mobile-scroll-container bpk-horizontal-nav bpk-horizontal-nav--show-light-underline"
    style="height: auto;"
  >
    <div
      class="bpk-mobile-scroll-container__scroller"
    >
      <nav
        class="bpk-mobile-scroll-container__inner"
      >
        <div
          class="bpk-horizontal-nav__list"
          role="tablist"
        >
          <div>
            <li
              type="light"
            >
              One
            </li>
          </div>
          <div>
            <li
              type="light"
            >
              Two
            </li>
          </div>
          <div>
            <li
              type="light"
            >
              Three
            </li>
          </div>
        </div>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BpkHorizontalNav should render correctly with arbitrary props 1`] = `
<DocumentFragment>
  <div
    class="bpk-mobile-scroll-container bpk-horizontal-nav bpk-horizontal-nav--show-default-underline"
    data-arbitrary-1="arbirary-value-1"
    data-arbitrary-2="arbirary-value-2"
    style="height: auto;"
  >
    <div
      class="bpk-mobile-scroll-container__scroller"
    >
      <nav
        class="bpk-mobile-scroll-container__inner"
      >
        <div
          class="bpk-horizontal-nav__list"
          role="tablist"
        >
          My nav content.
        </div>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BpkHorizontalNav should render correctly with custom "className" prop 1`] = `
<DocumentFragment>
  <div
    class="bpk-mobile-scroll-container bpk-horizontal-nav bpk-horizontal-nav--show-default-underline my-custom-class-name"
    style="height: auto;"
  >
    <div
      class="bpk-mobile-scroll-container__scroller"
    >
      <nav
        class="bpk-mobile-scroll-container__inner"
      >
        <div
          class="bpk-horizontal-nav__list"
          role="tablist"
        >
          My nav content.
        </div>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BpkHorizontalNav should render correctly with custom "leadingScrollIndicatorClassName" prop 1`] = `
<DocumentFragment>
  <div
    class="bpk-mobile-scroll-container bpk-horizontal-nav bpk-horizontal-nav--show-default-underline"
    style="height: auto;"
  >
    <div
      class="bpk-mobile-scroll-container__scroller"
    >
      <nav
        class="bpk-mobile-scroll-container__inner"
      >
        <div
          class="bpk-horizontal-nav__list"
          role="tablist"
        >
          My nav content.
        </div>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BpkHorizontalNav should render correctly with custom "trailingScrollIndicatorClassName" prop 1`] = `
<DocumentFragment>
  <div
    class="bpk-mobile-scroll-container bpk-horizontal-nav bpk-horizontal-nav--show-default-underline"
    style="height: auto;"
  >
    <div
      class="bpk-mobile-scroll-container__scroller"
    >
      <nav
        class="bpk-mobile-scroll-container__inner"
      >
        <div
          class="bpk-horizontal-nav__list"
          role="tablist"
        >
          My nav content.
        </div>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;
